Payments getall (#412)
MewsSystems/gitbook-connector-api: operations/payments.md

#### Payment next action discriminator

* `RedirectToUrl` - Redirect customer to a URL where they can complete their payment.

## Get all payments

> Returns all payments. Note this operation uses [Pagination](../guidelines/pagination.md).

Reviewer: I think we could add a fuller description than this. If it was simply 'get all payments' then there would be no parameters. In fact PaymentIds, BillIds and AccountingStates are **required** parameters, so as a minimum it should be something like "Returns all payments with specified accounting states belonging to specified bills". PaymentIds is required - is that correct, does that mean it only returns specified payments?

Author: If you can come up with something would be good, I can't come up with something good, since it's getting all the payments (I've corrected optional/required filters).

                        }
                    ]
                }
            },
            "OriginalAmount": {
                "Currency": "EUR",
                "NetValue": -300.00,
                "GrossValue": -300.00,
                "TaxValues": [],
                "Breakdown": {
                    "Items": [
                        {
                            "TaxRateCode": null,
                            "NetValue": -300.00,
                            "TaxValue": 0.0
                        }
                    ]
                }
            },
            "Notes": null,
            "SettlementId": null,
            "ConsumedUtc": "2023-03-02T12:12:35Z",
            "ClosedUtc": "2023-03-02T12:12:35Z",
            "ChargedUtc": "2023-03-02T12:12:32Z",
            "CreatedUtc": "2023-03-02T12:12:32Z",
            "UpdatedUtc": "2023-03-02T12:12:37Z",
            "AccountingState": "Closed",
            "State": "Charged",
            "Identifier": "",
            "PaymentType": "Cash",
            "DataDiscriminator": null,
            "Data": null
        }
    ],
    "Cursor": "d98c9611-0006-4691-a835-af2e00b170c4"
}
```

> 

Reviewer: There is no definition for the response. I know it's pretty trivial, but for consistency there should be a table defining the response, consisting of the "Payments" array of Payment objects, plus "Cursor" for pagination.

#### Payment

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Id` | string | required | Unique identifier of the payment. |
| `AccountId` | string | required | Unique identifier of the account (for example [Customer](customers.md#customer)) the payment belongs to. |
| `BillId` | string | optional | Unique identifier of the [Bill](bills.md#bill) the payment is assigned to. |
| `AccountingCategoryId` | string | optional | Unique identifier of the [Accounting category](accountingcategories.md#accounting-category) the payment belongs to. |

> | `Amount` | [Amount value](#amount-value) | required | Payment's amount, negative amount represents either rebate or a payment. |

Reviewer: Not sure I understand the description related to negative amounts "Payment's amount, negative amount represents either rebate or a payment." Remove 'either rebate or' and it reads like a payment is a payment?

Author: I've taken this description from `accountingItems` - `payment item` since this endpoint is basically taken from `accounting items` .

> | `OriginalAmount` | [Amount value](#amount-value) | required | Payment's original amount, negative amount represents either rebate or a payment. Contains the earliest known value in conversion chain. |

Reviewer: Same comment as above.

| `Notes` | string | optional | Additional notes. |
| `SettlementId` | string | optional | Identifier of the settled payment from the external system (ApplePay/GooglePay). | 
| `ConsumedUtc` | string | required | Date and time of the item consumption in UTC timezone in ISO 8601 format. |
| `ClosedUtc` | string | optional | Date and time of the payment bill closure in UTC timezone in ISO 8601 format. |
| `ChargedUtc` | string | required | Charged date and time of the payment in UTC timezone in ISO 8601 format. |
| `CreatedUtc` | string | required | Creation date and time of the payment created in UTC timezone in ISO 8601 format. |
| `UpdatedUtc` | string | required | Last update date and time of the payment in UTC timezone in ISO 8601 format. |
| `AccountingState` | string [Accounting item state](#accounting-item-state) | required | Accounting state of the payment. |
| `State` | string [Payment state](#payment-state) | required | Payment state of the payment. |
| `Identifier` | string | optional | Additional unique identifier of the payment dependent on the payment type. |
| `PaymentType` | string [Payment type](#payment-type) | required | Payment state of the payment. |

> | `DataDiscriminator` | object [Payment item data](#payment-data-discriminator) | required | Discriminator pointing to the field in the [Data](#payment-data-discriminator) field that contains additional information. |

Reviewer: Is the definition of 'Payment data discriminator' missing?

Author: It's not, it's at the bottom or you mean something specific?

| `Data` | object [Payment data](#payment-data) | required | Additional payment data. |

#### Payment data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `CardData` | object [Card data](#payment-data)| optional | Contains additional data for card payment. |

Reviewer: ```suggestion
| `CardData` | object [Card data](#card-data)| optional | Contains additional data in the case of a card payment. |
```

> | `InvoiceData` | object [Invoice data](#payment-data) | optional | Contains additional data for invoice payment. |

Reviewer: ```suggestion
| `InvoiceData` | object [Invoice data](#invoice-data) | optional | Contains additional data in the case of an invoice payment. |
```

#### Card data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `PaymentCardId` | string | optional | Unique identifier of the payment card. |

#### Invoice data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `PaymentCardId` | string | optional | Unique identifier of the invoice. |

#### Time interval

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `StartUtc` | string | required | Start of the interval in UTC timezone in ISO 8601 format. |
| `EndUtc` | string | required | End of the interval in UTC timezone in ISO 8601 format. |

#### Accounting item state

* `Open` - Accounting items which carry a non-zero value, are open, and have not been closed on a bill or invoice.
* `Closed` - Accounting items which carry a non-zero value and have been closed on a bill or invoice.
* `Inactive` - Accounting items which are either of zero value and have not been canceled, if the state of the payment item is Pending or Failed, or items of optional reservations. Until the reservation is confirmed, all its accounting items are Inactive.

> * `Canceled` - Accounting items which have been canceled, regardless of whether the item is of zero value

Reviewer: ```suggestion
* `Canceled` - Accounting items which have been canceled, regardless of whether the item is of zero value.
```

#### Card data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `PaymentCardId` | string | optional | Unique identifier of the payment card. |

#### Invoice data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `PaymentCardId` | string | optional | Unique identifier of the invoice. |

#### Time interval

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `StartUtc` | string | required | Start of the interval in UTC timezone in ISO 8601 format. |
| `EndUtc` | string | required | End of the interval in UTC timezone in ISO 8601 format. |

#### Accounting item state

* `Open` - Accounting items which carry a non-zero value, are open, and have not been closed on a bill or invoice.
* `Closed` - Accounting items which carry a non-zero value and have been closed on a bill or invoice.
* `Inactive` - Accounting items which are either of zero value and have not been canceled, if the state of the payment item is Pending or Failed, or items of optional reservations. Until the reservation is confirmed, all its accounting items are Inactive.

> * `Canceled` - Accounting items which have been canceled, regardless of whether the item is of zero value.

Reviewer 2: ```suggestion
* `Canceled` - Accounting items which have been canceled, regardless of whether the item is of zero value.
* ,,,
```

#### Payment state

* `Charged`
* `Canceled`
* `Pending`
* `Failed`
* `Verifying`

#### Payment type

* `CreditCard`
* `Invoice`
* `Cash`
* `Unspecified`
* `BadDebts`
* `WireTransfer`
* `ExchangeRateDifference`
* `ExchangeRoundingDifference`
* `BankCharges`
* `Cheque`

> * `Other`

Reviewer 2: One of the main reasons why we needed the endpoint was to add the support for more payment types. as this was a strict enum in the accountingItems it can't be extended because it is a breaking change. Also some of the types can be added right away.
```suggestion
* `Other`
* ...
```

| `Notes` | string | optional | Additional notes. |
| `SettlementId` | string | optional | Identifier of the settled payment from the external system (ApplePay/GooglePay). | 
| `ConsumedUtc` | string | optional | Date and time of the item consumption in UTC timezone in ISO 8601 format. |
| `ClosedUtc` | string | optional | Date and time of the payment bill closure in UTC timezone in ISO 8601 format. |
| `ChargedUtc` | string | optional | Charged date and time of the payment in UTC timezone in ISO 8601 format. |
| `CreatedUtc` | string | required | Creation date and time of the payment created in UTC timezone in ISO 8601 format. |
| `UpdatedUtc` | string | required | Last update date and time of the payment in UTC timezone in ISO 8601 format. |
| `AccountingState` | string [Accounting item state](#accounting-item-state) | required | Accounting state of the payment. |
| `State` | string [Payment state](#payment-state) | required | Payment state of the payment. |
| `Identifier` | string | optional | Additional unique identifier of the payment. |
| `PaymentType` | string [Payment type](#payment-type) | required | Payment state of the payment. |

> | `DataDiscriminator` | object [Payment item data](#payment-data-discriminator) | optional | Discriminator pointing to the field in the [Data](#payment-data) field that contains additional information. |

Reviewer: This section looks wrong to me. It doesn't work the same way as other instances of data discriminators in the API. Please confirm, but I think it should work like this:

| `DataDiscriminator` | string [Payment data discriminator](#payment-data-discriminator) | optional | Determines the type of additional payment data. |
| `Data` | object | optional | Additional payment data. Object structure depends on the value of `DataDiscriminator`. |

| `Data` | object [Payment data](#payment-data) | optional | Additional payment data. |

> 

Reviewer: I would put the discriminator definitions right underneath here ->

#### Payment data discriminator

* `CardData` - Data related to payment card
* `InvoiceData` - Data related to invoice
* ...

And then finally I would put the definitions of the two types of data, called "CardData payment data" and "InvoiceData payment data".

Author: I've move it but the definition is up for discussion, about whole data, comment above.

> #### Payment data

Reviewer: If this works like other instances of data discriminators in the API, which I assume it does, then "Payment data" here isn't a real JSON object and it can be removed.

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `CardData` | object [Card data](#card-data)| optional | Contains additional data in the case of a card payment. |
| `InvoiceData` | object [Invoice data](#invoice-data) | optional | Contains additional data in the case of an invoice payment. |

> #### Card data

Reviewer: ```suggestion
#### CardData payment data
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `PaymentCardId` | string | optional | Unique identifier of the payment card. |

> #### Invoice data

Reviewer: ```suggestion
#### InvoiceData payment data
```

#### Time interval

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `StartUtc` | string | required | Start of the interval in UTC timezone in ISO 8601 format. |
| `EndUtc` | string | required | End of the interval in UTC timezone in ISO 8601 format. |

#### Accounting item state

* `Open` - Accounting items which carry a non-zero value, are open, and have not been closed on a bill or invoice.
* `Closed` - Accounting items which carry a non-zero value and have been closed on a bill or invoice.
* `Inactive` - Accounting items which are either of zero value and have not been canceled, if the state of the payment item is Pending or Failed, or items of optional reservations. Until the reservation is confirmed, all its accounting items are Inactive.
* `Canceled` - Accounting items which have been canceled, regardless of whether the item is of zero value.
* ...

#### Payment state

* `Charged`
* `Canceled`
* `Pending`
* `Failed`
* `Verifying`
* ...

#### Payment type

* `CreditCard`
* `Invoice`
* `Cash`
* `Unspecified`
* `BadDebts`
* `WireTransfer`
* `ExchangeRateDifference`
* `ExchangeRoundingDifference`
* `BankCharges`
* `Cheque`
* `Other`
* ...

> #### Payment data discriminator

Reviewer: I have suggested moving this definition higher up - see my comments earlier.

                    ]
                }
            },
            "OriginalAmount": {
                "Currency": "GBP",
                "NetValue": -3700.0,
                "GrossValue": -3700.0,
                "TaxValues": [],
                "Breakdown": {
                    "Items": [
                        {
                            "TaxRateCode": null,
                            "NetValue": -3700.0,
                            "TaxValue": 0.0
                        }
                    ]
                }
            },
            "Notes": null,
            "SettlementId": null,
            "ClosedUtc": null,
            "ConsumedUtc": "2023-03-02T12:12:35Z",
            "ChargedUtc": "2023-03-06T07:31:52Z",
            "CreatedUtc": "2023-03-06T07:31:51Z",
            "UpdatedUtc": "2023-03-06T07:31:53Z",
            "AccountingState": "Open",
            "State": "Charged",
            "Identifier": "ch_764309db-4bcd-4f2c-ad6a-1c178089deec",
            "PaymentType": "CreditCardPayment",
            "Data": {
                "Discriminator": "CreditCard",
                "CreditCard": {
                    "CreditCardId": "c922266b-291d-4e25-9df7-afbd007c1991",

>                     "Transaction": null,

Reviewer: ```suggestion
                    "Transaction": null
```

| `CreditCard` | object [Credit card data](#card-data)| optional | Contains additional data in the case of a card payment. |
| `Invoice` | object [Invoice data](#invoice-data) | optional | Contains additional data in the case of an invoice payment. |
| `External` | object [External data](#external-data) | optional | Contains additional data in the case of an external payment. |

#### Payment data discriminator

* `CreditCard`
* `Invoice`
* `External`
* ...

#### CreditCard payment data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `CreditCardId` | string | optional | Unique identifier of the payment card. |
| `Transaction` | object [Credit card transaction](#credit-card-transaction) | optional | The credit card payment transactions. |

#### Invoice payment data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `InvoiceId` | string | optional | Unique identifier of the invoice [Bill](bills.md#bill). |

#### External payment data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `Type` | string [External payment data](#external-payment-type) | required | Type of the external payment. *Except for the enterprises based in the French Legal Environment. Unspecified is considered as fraud. |

Reviewer: ```suggestion
| `Type` | string [External payment type](#external-payment-type) | required | Type of the external payment. *Except for enterprises based in the French Legal Environment. `Unspecified` is considered as fraud. |
```
